specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp

```diff
@@ -0,0 +1,202 @@
+  content: ChatMessageContent;
+
+  @doc("An optional name for the participant.")
+  name?: string;
+}
+
+@added(ServiceApiVersions.v2023_05_15)
+@doc("A request chat message representing response or action from the assistant.")
+model ChatRequestAssistantMessage extends ChatRequestMessage {
+  @doc("The chat role associated with this message, which is always 'assistant' for assistant messages.")
+  role: ChatRole.assistant;
+
+  #suppress "@azure-tools/typespec-azure-core/no-nullable" "explicitly nullable in mirrored API"
+  @doc("The content of the message.")
+  content: string | null;
+
+  @doc("An optional name for the participant.")
+  name?: string;
+
+  @projectedName("json", "tool_calls")
+  @added(ServiceApiVersions.v2023_12_01_Preview)
+  @doc("""
+  The tool calls that must be resolved and have their outputs appended to subsequent input messages for the chat
+  completions request to resolve as configured.
+  """)
+  toolCalls?: ChatCompletionsToolCall[];
+
+  @added(ServiceApiVersions.v2023_07_01_Preview)
+  @removed(ServiceApiVersions.v2023_12_01_Preview)
+  @projectedName("json", "function_call")
+  @doc("""
+  The function call that must be resolved and have its output appended to subsequent input messages for the chat
+  completions request to resolve as configured.
+  """)
+  functionCall?: FunctionCall;
+}
+
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("A request chat message representing requested output from a configured tool.")
+model ChatRequestToolMessage extends ChatRequestMessage {
+  @doc("The chat role associated with this message, which is always 'tool' for tool messages.")
+  role: ChatRole.tool;
+
+  #suppress "@azure-tools/typespec-azure-core/no-nullable" "explicitly nullable in mirrored API"
+  @doc("The content of the message.")
+  content: string | null;
+
+  @projectedName("json", "tool_call_id")
+  @doc("The ID of the tool call resolved by the provided content.")
+  toolCallId: string;
+}
+
+@added(ServiceApiVersions.v2023_07_01_Preview)
+@removed(ServiceApiVersions.v2023_12_01_Preview)
+@doc("A request chat message representing requested output from a configured function.")
+model ChatRequestFunctionMessage extends ChatRequestMessage {
+  @doc("The chat role associated with this message, which is always 'function' for function messages.")
+  role: ChatRole.function;
+
+  @doc("The name of the function that was called to produce output.")
+  name: string;
+
+  @doc("The output of the function as requested by the function call.")
+  content: string;
+}
+
+@added(ServiceApiVersions.v2023_05_15)
+@doc("A representation of a chat message as received in a response.")
+model ChatResponseMessage {
+  @doc("The chat role associated with the message.")
+  role: ChatRole;
+
+  #suppress "@azure-tools/typespec-azure-core/no-nullable" "explicitly nullable in mirrored API"
+  @doc("The content of the message.")
+  content: string | null;
```

Comment:
super nit: In the `streaming|non-streaming` model separation PR I extracted `alias NullableFloat = float32 | null;` into `common.tsp` . Seeing this, made me think we should probably do the same. Maybe extract this into a `.tsp` file of its own with all these nullable types.

Reply:
Messing a bit in the TSP playground, something like:

```ts
alias NullableField<T> = T | null;

model Dog {
  name: NullableField<string>;
}
```

Yields the following OpenAPIv3 definition:

```yaml
openapi: 3.0.0
info:
  title: (title)
  version: 0000-00-00
tags: []
paths: {}
components:
  schemas:
    Dog:
      type: object
      required:
        - name
      properties:
        name:
          type: string
          nullable: true
```

Totally aware that now is not the time for such housekeeping though 😄

Reply:
Yeah, and I think the generics mess with emitters in funny ways. Let's do this all at once (ideally together with the change you're working on).